First version

gavin-peng opened the pull request.

gavin-peng commented:
see ticket https://jira.oicr.on.ca/browse/GRD-833

pruzanov reviewed (changes requested):
Since this is a first implementation I would strongly recommend to avoid using separate wdl files. It will add additional maintenance headaches, especially given that the files are almost identical. I see that in this case we could introduce a **mode** variable which would route the execution two one of two tasks - one for PERL and another for R mode. We had a similar approach implemented in bclconvert wf which allows choosing between hpc and dragen modes.

lheisler commented:
I'm also inclined to choose and implement just one of the versions, perl or java.  I'm working with CHUM now to get their vardict output and we can run on a test sample to see what each version generates.   I'll also investigate a bit more to better understand if it is clear that a specific version is being used in the genpipes pipelines that CHUM is using

gavin-peng commented:
Yes we are in the process of choosing one of perl or java version. For a small inputs they generate indentical output, the problem is for large inputs they both have outof memory issue. I have tested java multiple times increasing memory, last attempt used 512G job memory with > 300G inputs from CHUM, still failed because of memory after 60h 32m. 
The vardict github page claims java version is 10X faster, for now I haven't observed huge difference in speed.

gavin-peng commented:
> Since this is a first implementation I would strongly recommend to avoid using separate wdl files. It will add additional maintenance headaches, especially given that the files are almost identical. I see that in this case we could introduce a **mode** variable which would route the execution two one of two tasks - one for PERL and another for R mode. We had a similar approach implemented in bclconvert wf which allows choosing between hpc and dragen modes.

The perl version now removed

pruzanov commented on tests/calculate.sh:
```diff
@@ -0,0 +1,8 @@
+
+cd $1
+
+find . -name '*.vcf'  | xargs md5sum 
```
I would recommend omitting header lines: ` for f in $(find . -xtype f -name "*.vcf.gz" | sort -V);do zcat $f | grep -v ^# | md5sum`

gavin-peng replied:
changed as your suggested, thanks!

pruzanov reviewed (approved):
I approved, but recommend modifying _calculate.sh_ test script

lheisler commented on vardict.wdl:
```diff
@@ -168,7 +175,7 @@ task runVardict {
         String READ_POSTION_FILTER = 5
```
READ_POSITION_FILTER

gavin-peng replied:
fixed. thanks!

lheisler commented:
include the parameter
-th 8
which will give opportunity for the software to use multiple slots on the node

i tested this on one small dataset and it reduced the processing time from
166m to 33m

lheisler commented:
even with threading, this runs very long.

for vardict, and for the neoantigen pipeline in particular, we should be okay to exclude intergenic regions

I tested this out with a bedfile from UCSC knowngenes, splitting the bed file by chromosome, and a simplified command.
jobs completed in the range of 3 m to 1258 m (20 h)

test command was
varDict -G REF -N MoHQ-CM-1-180.DT -b 'Tbam|Nbam' -c 1 -S 2 -E 3 -g 4 -th 8 -D -y knownGene.chr.bed
(the above is extracted from the current workflow command, which pipes to additional filtering commands. i removed the pipes for testing purposes)

my tests were with knownGene , but i think we can use the same bed file that is being used for TMB assessment


https://bitbucket.oicr.on.ca/projects/GSI/repos/interval-files/browse/accredited/MANE_Select_v1.3.bed

gavin-peng commented:
Trying to run vardict with -th 8 and uses UCSC knowngene bed file, seeing speed much faster, chrY completed in 28 minutes, versus 33h 22m last time. Still running though.

lheisler commented:
I had noted on the ticket
https://jira.oicr.on.ca/browse/GRD-833

that the bed file should be reduced to just the merged intervals;  the knownGene.bed originally supplied contained intervals for each gene and isoforms, and there was much overlap between intervals.

lheisler commented:
the final merged output is limited to ONLY chrY

This workflow parallelizes by chromosome, and i checked my run; each of the chromosomes are being processed.
The problem must be at the merging step.   
I think you are looping through files, and redirecting to an output file with >, and overwriting, when you should be appending >>

gavin-peng commented:
@lheisler The resulting merged vcf does have problem, but the cause is more subtle. It's not caused by "edirecting to an output file with >", because spliting chromosome is done on task runVardict, which after parrallezing generated vcf files for every chromosomes, the array of vcfs as inputs to mergeVcfs task, does not need redirect, the command is:
` java "-Xmx~{memory-3}g" -jar $PICARD_ROOT/picard.jar MergeVcfs \
    $input_args \
    O=~{tumor_sample_name}.vardict.vcf.gz \
    SEQUENCE_DICTIONARY=~{refDict}`
    which would  correctly merge vcfs from all chromosomes, the problem is the bash script I wrote before merge vcf, which supposed to fix a problem of vardict using illegal notation of variants. Need to fix this.

lheisler commented:
> @lheisler The resulting merged vcf does have problem, but the cause is more subtle. It's not caused by "edirecting to an output file with >", because spliting chromosome is done on task runVardict, which after parrallezing generated vcf files for every chromosomes, the array of vcfs as inputs to mergeVcfs task, does not need redirect, the command is: ` java "-Xmx~{memory-3}g" -jar $PICARD_ROOT/picard.jar MergeVcfs \ $input_args \ O=~{tumor_sample_name}.vardict.vcf.gz \ SEQUENCE_DICTIONARY=~{refDict}` which would correctly merge vcfs from all chromosomes, the problem is the bash script I wrote before merge vcf, which supposed to fix a problem of vardict using illegal notation of variants. Need to fix this.

ok, i had reviewed quickly and seemed to get the correct results when i rewrote that line locally.    let me know when fixed and i'll test again for the PR

gavin-peng commented:
> > @lheisler The resulting merged vcf does have problem, but the cause is more subtle. It's not caused by "edirecting to an output file with >", because spliting chromosome is done on task runVardict, which after parrallezing generated vcf files for every chromosomes, the array of vcfs as inputs to mergeVcfs task, does not need redirect, the command is: ` java "-Xmx~{memory-3}g" -jar $PICARD_ROOT/picard.jar MergeVcfs \ $input_args \ O=~{tumor_sample_name}.vardict.vcf.gz \ SEQUENCE_DICTIONARY=~{refDict}` which would correctly merge vcfs from all chromosomes, the problem is the bash script I wrote before merge vcf, which supposed to fix a problem of vardict using illegal notation of variants. Need to fix this.
> 
> ok, i had reviewed quickly and seemed to get the correct results when i rewrote that line locally. let me know when fixed and i'll test again for the PR

I rewrote the bash script for normalizing vardict vcf in python and splited it inti separte task, tested in small inputs the issue solved. There is a run using your last inputs: http://cromwell-job-manager-dev.gsi.oicr.on.ca:4202/jobs/7cc63142-bf3b-4854-bc06-0c6a61bd377b?q=name%3Dvardict still running

lheisler commented:
The normalization task was designed to deal with issues identified in the vardict output associated with structural variants, that appear to be incompatible with picard.  these fail even with validation stringency turned to silent

it produces a modified alt string, for which Gavin had identified the code that produces it here:
https://github.com/AstraZeneca-NGS/VarDictJava/blob/master/src/main/java/com/astrazeneca/vardict/modules/StructuralVariantsProcessor.java#L1430

The normalization function removed the 
<dup" + (mlen - 2 * Configuration.SVFLANK) + ">
information from the ALT string, and then accounts for it by adding additional information in to the INFO fields to describe what it was removed

**i'm hesitant to include this in the workflow without understanding further, as it does modify the vardict output**

the current need for this workflow is for mutation calling as part of the neoantigen workflow.  We may consider using this caller more generally, and releasing the files that it produces.  Ideally these represent what vardict produces; if we are modifying the vardict output we do want to be certain that this is clearly documented.

We are **NOT** at this point utilizing Structural variant calls from this tool.   These are not considered by the neoantigen prediction pipeline.

We can turn off SV calling by adding the argument
-U|--nosv  Turn off structural variant calling

i tested this on the test CHUM sample, on a region of chromosome2 where the issue had been noted using the bed interval
chr2	11482340	11642788

this resulted in a reduction of calls in this region from 536 to 535, removing the single structural variant

----
Let's turn off SV calling in the workflow, and shelf this solution for later consideration should we want to run vardict for SV calling in the future.

this should allow us to take the scattered calls and join them together with picard MergeBam directly and not require modification to the records.

gavin-peng commented:
I did a test using --nosv option and skipped normaliztion task, the run succeeded mergeVcfs without complaining of format. The speed also likely faster. 6 hurs to 4 hurs. So I will moving forward with this.

lheisler commented:
Testing the revised version,

i immediately get an error in the first task, call-splitBedByChromosome
mkdir: cannot create directory ‘split_beds’: File exists

lheisler commented:
All good now, lets get this merged and tagged so we can run in stage
